File: apps/android/app/src/main/java/com/kris99/baozi/android/ui/terminal/GhosttyRendererBackendBridge.kt
package com.kris99.baozi.android.ui.terminal

import android.os.Handler
import android.os.Looper
import com.kris99.baozi.android.core.bridge.GhosttyRendererBridge
import uniffi.codex_mobile_client.TerminalCellMetrics
import uniffi.codex_mobile_client.TerminalCellRange
import uniffi.codex_mobile_client.TerminalKeyAction
import uniffi.codex_mobile_client.TerminalKeyCode
import uniffi.codex_mobile_client.TerminalKeyEvent
import uniffi.codex_mobile_client.TerminalKeyMods
import uniffi.codex_mobile_client.TerminalRendererBackend
import java.util.concurrent.CountDownLatch
import java.util.concurrent.atomic.AtomicReference

/**
 * Kotlin implementation of the Rust-defined `TerminalRendererBackend` callback
 * interface. The Rust [`uniffi.codex_mobile_client.TerminalRenderer`] tick task
 * invokes these methods from a tokio worker; we hop to the main thread before
 * touching the (non-thread-safe) Ghostty surface APIs.
 *
 * Selection state lives here because Ghostty's C surface doesn't expose a
 * public setter for the painted overlay — the platform paints handles
 * itself and uses the stored range to satisfy `readSelection` via
 * `nativeReadText`. The Compose overlay subscribes to
 * [`onSelectionRangeChanged`] to redraw when Rust pushes a new range.
 */
internal class GhosttyRendererBackendBridge(
    private val surface: GhosttyRendererBridge.GhosttyRendererSurface,
    private val onRequestRedraw: () -> Unit,
    /// Closure invoked when the renderer needs to push raw bytes to the
    /// PTY input direction (terminal → shell). Bracketed-paste payloads
    /// flow through here so they reach the running process unmodified.
    private val onPasteBytes: (ByteArray) -> Unit,
) : TerminalRendererBackend {

    private val mainHandler = Handler(Looper.getMainLooper())
    private val selectionRange = AtomicReference<TerminalCellRange?>(null)

    /// Main-thread callback fired whenever the stored selection range
    /// changes. The terminal surface installs this to drive handle
    /// repaints + ActionMode visibility.
    @Volatile
    var onSelectionRangeChanged: ((TerminalCellRange?) -> Unit)? = null

    override fun setFocus(focused: Boolean) {
        runOnMain { surface.setFocus(focused) }
    }

    override fun setOcclusion(occluded: Boolean) {
        runOnMain { surface.setOcclusion(occluded) }
    }

    override fun requestRedraw() {
        runOnMain { onRequestRedraw() }
    }

    override fun applyConfigFile(path: String) {
        runOnMainBlocking {
            surface.applyConfig(path)
            onRequestRedraw()
            true
        }
    }

    override fun dispatchKey(event: TerminalKeyEvent) {
        val action = when (event.action) {
            TerminalKeyAction.RELEASE -> 0
            TerminalKeyAction.PRESS -> 1
            TerminalKeyAction.REPEAT -> 2
        }
        val key = bridgeKey(event.code)
        val mods = packMods(event.mods)
        val text = event.text.ifEmpty { null }
        runOnMain { surface.sendKey(action, key, mods, text, composing = false) }
    }

    override fun dispatchText(text: String, composing: Boolean) {
        runOnMain {
            if (composing) {
                surface.sendPreedit(text.ifEmpty { null })
            } else if (text.isNotEmpty()) {
                surface.sendText(text)
            }
        }
    }

    override fun dispatchPaste(bytes: ByteArray) {
        // Bracketed-paste bytes must travel PTY-input direction (terminal
        // → shell), so we feed them through the same callback Ghostty's
        // `external_pty_write` triggers on key input. `surface.write`
        // would route them PTY-output direction (paint), which is wrong.
        runOnMain { onPasteBytes(bytes) }
    }

    override fun readSelection(): String? {
        val range = selectionRange.get() ?: return null
        return readTextBlocking(
            startRow = range.start.row.toInt(),
            startCol = range.start.col.toInt(),
            endRow = range.end.row.toInt(),
            endCol = range.end.col.toInt(),
        )
    }

    override fun readText(
        startRow: UInt,
        startCol: UInt,
        endRow: UInt,
        endCol: UInt,
    ): String? = readTextBlocking(
        startRow = startRow.toInt(),
        startCol = startCol.toInt(),
        endRow = endRow.toInt(),
        endCol = endCol.toInt(),
    )

    override fun cellMetrics(): TerminalCellMetrics {
        val size = runOnMainBlocking { surface.surfaceSize() }
        if (size == null) {
            return TerminalCellMetrics(
                cellWidthPx = 0f,
                cellHeightPx = 0f,
                cols = 0u,
                rows = 0u,
                viewportTop = 0u,
            )
        }
        return TerminalCellMetrics(
            cellWidthPx = size.cellWidthPx.toFloat(),
            cellHeightPx = size.cellHeightPx.toFloat(),
            cols = size.columns.toUInt(),
            rows = size.rows.toUInt(),
            // Selection coords are viewport-relative; scrollback gating
            // is the OSC parser's job, not ours.
            viewportTop = 0u,
        )
    }

    override fun setSelectionOverlay(range: TerminalCellRange?) {
        selectionRange.set(range)
        val callback = onSelectionRangeChanged
        runOnMain { callback?.invoke(range) }
    }

    /// Snapshot the current selection range (for the overlay view / edit
    /// menu without going through Rust).
    fun currentSelectionRange(): TerminalCellRange? = selectionRange.get()

    private fun readTextBlocking(
        startRow: Int,
        startCol: Int,
        endRow: Int,
        endCol: Int,
    ): String? = runOnMainBlocking {
        surface.readText(startRow, startCol, endRow, endCol)
    }

    private fun <T> runOnMainBlocking(block: () -> T?): T? {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            return block()
        }
        val result = AtomicReference<T?>(null)
        val latch = CountDownLatch(1)
        mainHandler.post {
            try {
                result.set(block())
            } finally {
                latch.countDown()
            }
        }
        latch.await()
        return result.get()
    }

    private fun packMods(mods: TerminalKeyMods): Int {
        var bits = 0
        if (mods.shift) bits = bits or (1 shl 0)
        if (mods.ctrl) bits = bits or (1 shl 1)
        if (mods.alt) bits = bits or (1 shl 2)
        if (mods.meta) bits = bits or (1 shl 3)
        return bits
    }

    // Mirrors `BaoziBridgeKey` in ghostty_jni.cpp; the JNI bridge does the
    // final translation to ghostty_input_key_e.
    private fun bridgeKey(code: TerminalKeyCode): Int = when (code) {
        is TerminalKeyCode.Enter -> 1
        is TerminalKeyCode.Tab -> 2
        is TerminalKeyCode.Backspace -> 3
        is TerminalKeyCode.Escape -> 4
        is TerminalKeyCode.Space -> 5
        is TerminalKeyCode.ArrowUp -> 6
        is TerminalKeyCode.ArrowDown -> 7
        is TerminalKeyCode.ArrowLeft -> 8
        is TerminalKeyCode.ArrowRight -> 9
        is TerminalKeyCode.PageUp -> 10
        is TerminalKeyCode.PageDown -> 11
        is TerminalKeyCode.Home -> 12
        is TerminalKeyCode.End -> 13
        is TerminalKeyCode.Delete -> 14
        is TerminalKeyCode.Insert -> 15
        else -> 0
    }

    private fun runOnMain(block: () -> Unit) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            block()
        } else {
            mainHandler.post(block)
        }
    }
}

File: apps/ios/Sources/Baozi/Bridge/GhosttyRendererBackendBridge.swift
import Foundation

/// Thin Swift implementation of the Rust-defined `TerminalRendererBackend`
/// callback interface. Holds a weak reference to the platform-side
/// `BaoziGhosttyTerminal` and hops every Ghostty C call onto the main
/// thread (Ghostty's surface APIs are not thread-safe). The Rust tick task
/// invokes these methods on the shared tokio runtime.
///
/// Selection state lives here because Ghostty's C surface doesn't expose a
/// public setter for the painted selection range — the platform paints the
/// overlay itself and uses the stored range to satisfy `readSelection` via
/// `ghostty_surface_read_text`. The UI overlay view subscribes to
/// `onSelectionRangeChanged` to redraw handles when Rust pushes a new range.
final class GhosttyRendererBackendBridge: TerminalRendererBackend, @unchecked Sendable {
    private weak var terminal: BaoziGhosttyTerminal?

    /// Most recently pushed selection range (viewport-relative). `nil` when
    /// no selection is active. Written from the Rust runtime via
    /// `setSelectionOverlay` and read from the main thread by the overlay
    /// view + edit menu. Guarded by `selectionLock` to keep the write/read
    /// race safe — the storage is a single optional, no fancy state.
    private let selectionLock = NSLock()
    private var selectionRange: TerminalCellRange?

    /// Callback fired on the main thread whenever the stored selection
    /// range changes. The terminal view installs this to drive handle
    /// repaints + edit-menu visibility.
    var onSelectionRangeChanged: ((TerminalCellRange?) -> Void)?

    init(terminal: BaoziGhosttyTerminal) {
        self.terminal = terminal
    }

    func setFocus(focused: Bool) {
        let terminal = self.terminal
        DispatchQueue.main.async {
            terminal?.setFocused(focused)
        }
    }

    func setOcclusion(occluded: Bool) {
        let terminal = self.terminal
        DispatchQueue.main.async {
            terminal?.setOcclusion(occluded)
        }
    }

    func requestRedraw() {
        // UIKit Ghostty surfaces render through Ghostty's own renderer thread.
        // Ghostty's wakeup callback drains the app mailbox; this Rust-side
        // renderer callback only exists for Android's app-thread EGL path.
    }

    func applyConfigFile(path: String) {
        let terminal = self.terminal
        if Thread.isMainThread {
            try? terminal?.applyConfig(atPath: path)
        } else {
            DispatchQueue.main.async {
                try? terminal?.applyConfig(atPath: path)
            }
        }
    }

    func dispatchKey(event: TerminalKeyEvent) {
        let terminal = self.terminal
        let action = Int32(GhosttyKeyTranslator.action(for: event.action))
        let baoziKey = GhosttyKeyTranslator.baoziKey(for: event.code)
        let mods = Int32(GhosttyKeyTranslator.mods(for: event.mods))
        let text = event.text.isEmpty ? nil : event.text
        DispatchQueue.main.async {
            _ = terminal?.dispatchKeyAction(
                action,
                key: baoziKey,
                mods: mods,
                text: text,
                composing: false
            )
        }
    }

    func dispatchText(text: String, composing: Bool) {
        let terminal = self.terminal
        DispatchQueue.main.async {
            if composing {
                terminal?.setPreeditText(text.isEmpty ? nil : text)
            } else {
                terminal?.sendText(text)
            }
        }
    }

    func dispatchPaste(bytes: Data) {
        // Bracketed-paste bytes must travel PTY-input direction
        // (terminal → shell), not PTY-output direction. The terminal's
        // `inputHandler` is the same closure Ghostty's
        // `external_pty_write` ultimately fires when the user types, so
        // we reuse it: the platform-side controller forwards the bytes
        // to the running process unchanged. Writing them through
        // `writeOutput` would paint the wrapper on screen instead.
        let terminal = self.terminal
        DispatchQueue.main.async {
            terminal?.inputHandler?(bytes)
        }
    }

    func readSelection() -> String? {
        let range = currentSelectionRange()
        guard let range else { return nil }
        // `read_text` must run on the same thread as other Ghostty surface
        // calls. We're invoked from the Rust tick task, so hop to main and
        // block long enough to return — bounded waits keep this safe under
        // a misbehaving renderer (no deadlock with the renderer's tokio
        // runtime because no main-thread caller is waiting on us).
        return runOnMainBlocking { [weak terminal] in
            terminal?.readText(
                fromRow: range.start.row,
                column: range.start.col,
                toRow: range.end.row,
                column: range.end.col
            )
        }
    }

    func readText(startRow: UInt32, startCol: UInt32, endRow: UInt32, endCol: UInt32) -> String? {
        runOnMainBlocking { [weak terminal] in
            terminal?.readText(
                fromRow: startRow,
                column: startCol,
                toRow: endRow,
                column: endCol
            )
        }
    }

    func cellMetrics() -> TerminalCellMetrics {
        let metrics = runOnMainBlocking { [weak terminal] in
            terminal?.surfaceMetrics()
        } ?? BaoziGhosttySurfaceMetrics()
        return TerminalCellMetrics(
            cellWidthPx: Float(metrics.cellWidthPx),
            cellHeightPx: Float(metrics.cellHeightPx),
            cols: UInt32(metrics.columns),
            rows: UInt32(metrics.rows),
            // Viewport-relative selection: top-left of the visible area is
            // always row 0 in our coordinate system. Scrollback rows live
            // outside the viewport and aren't selectable through long-press
            // yet — the OSC parser's absolute-row tracking is separate.
            viewportTop: 0
        )
    }

    func setSelectionOverlay(range: TerminalCellRange?) {
        selectionLock.lock()
        selectionRange = range
        selectionLock.unlock()
        let callback = onSelectionRangeChanged
        DispatchQueue.main.async {
            callback?(range)
        }
    }

    /// Snapshot the current selection range. Used by `readSelection` and
    /// by the overlay view via `currentRange` to repaint.
    func currentSelectionRange() -> TerminalCellRange? {
        selectionLock.lock()
        defer { selectionLock.unlock() }
        return selectionRange
    }

    /// Run `work` on the main thread synchronously, returning its result.
    /// If we're already on main, runs inline; otherwise dispatches and
    /// waits. `DispatchQueue.main.sync` from a background thread is fine
    /// here because the Rust tick task never holds a lock the main thread
    /// could be waiting on.
    private func runOnMainBlocking<T>(_ work: @MainActor @Sendable () -> T) -> T {
        if Thread.isMainThread {
            return MainActor.assumeIsolated { work() }
        }
        return DispatchQueue.main.sync {
            MainActor.assumeIsolated { work() }
        }
    }
}

/// Translation: Rust `TerminalKey*` → bridge-level `BaoziGhosttyKey`.
/// Bridge does the final Ghostty-enum mapping in Obj-C.
enum GhosttyKeyTranslator {
    static func action(for value: TerminalKeyAction) -> Int {
        switch value {
        case .release: return 0
        case .press: return 1
        case .repeat: return 2
        }
    }

    static func mods(for value: TerminalKeyMods) -> Int {
        var bits = 0
        if value.shift { bits |= 1 << 0 }
        if value.ctrl { bits |= 1 << 1 }
        if value.alt { bits |= 1 << 2 }
        if value.meta { bits |= 1 << 3 }
        return bits
    }

    static func baoziKey(for value: TerminalKeyCode) -> BaoziGhosttyKey {
        switch value {
        case .enter: return .enter
        case .tab: return .tab
        case .backspace: return .backspace
        case .escape: return .escape
        case .space: return .space
        case .arrowUp: return .arrowUp
        case .arrowDown: return .arrowDown
        case .arrowLeft: return .arrowLeft
        case .arrowRight: return .arrowRight
        case .pageUp: return .pageUp
        case .pageDown: return .pageDown
        case .home: return .home
        case .end: return .end
        case .delete: return .delete
        case .insert: return .insert
        default: return .unidentified
        }
    }
}
